meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java

```diff
@@ -56,56 +69,81 @@ public StoreType getStoreType() {
   //  private RetryProtocol _retryProtocol;
 
 
-  private MetaClientConfig(String connectionAddress, long connectionTimeout, boolean enableAuth,
-      StoreType storeType) {
+  protected MetaClientConfig(String connectionAddress, long connectionInitTimeout,
+      long sessionTimeout, boolean enableAuth, StoreType storeType) {
     _connectionAddress = connectionAddress;
-    _connectionTimeout = connectionTimeout;
+    _connectionInitTimeout = connectionInitTimeout;
+    _sessionTimeout = sessionTimeout;
     _enableAuth = enableAuth;
     _storeType = storeType;
   }
 
-  public static class Builder {
-    private String _connectionAddress;
-
-    private long _connectionTimeout;
-    private boolean _enableAuth;
-    //private RetryProtocol _retryProtocol;
-    private StoreType _storeType;
+  public static class MetaClientConfigBuilder<B extends MetaClientConfigBuilder<B>> {
```

Comment:
Could you please clarify why this generics is necessary? Thanks.

Reply:
If we are not using generics with builder, calling setter function defined in base class will returns a bace class object instead.

```
ZkMetaClientConfigBuilder.setAuthEnabled(true).setZkSerializer(serializer)
```
will throw exception. 

https://stackoverflow.com/questions/17164375/subclassing-a-java-builder-class

Reply:
Thanks, that makes sense. This seems to be a very good use of generics. 
Can we make this class standalone? It's a `public static` anyway, that way might be more clear.